### ISSUE TITLE
try to understand YOLO source code for further modification

### ISSUE THREAD

**Issue by frabob2017** [ultralytics/ultralytics#1220]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello YOLO expert, it takes me a lot of time to understand your source code (YOLOv5). Sorry to bring this YOLOv5 question in V8 forum. I want to replace nn.conv2d with nn.conv3d for medical imaging detection since 2d version can not take advantage of the consecutive slice information. 

In the yolo.py file, function def parse_model(d, ch): line 317, I saw
if m in {
Conv, GhostConv, Bottleneck, GhostBottleneck, SPP, SPPF, DWConv, MixConv2d, Focus, CrossConv,
BottleneckCSP, C3, C3TR, C3SPP, C3Ghost, nn.ConvTranspose2d, DWConvTranspose2d, C3x}:

But from yolo5.yaml file, I only found module has Conv,C3,SPPF, nn.Upsample, Concat, Detect. May I ask the other modules defined in common.py are actually not used in YOLOV5 such as C3SPP, C3TR,C3Ghost, Focus, am I correct?  

I also need to modify loss function because I want to sum the loss from continuous slices loss  together instead of using the loss from a single slice after I change nn.conv3d.  It looks it needs a lot of coding. 

I also saw see a lot of module classes are not defined forward method. Does it further prove that these modules are not used in YOLOV5? I want to understand your code and hope to I can do some modification for my application.
![image](https://user-images.githubusercontent.com/57141860/222185388-7e4ab2d3-ea52-463b-928d-a9732036ee18.png)

This is a difficult task for me and hope that in the future, YOLO team can implement this feature.

**Comment by frabob2017** [comment]
I still have several questions to ask about source code:
1.   in yolov5.yaml file,   # [from, number, module, args]
  [[-1, 1, Conv, [64, 6, 2, 2]],  # 0-P1/2
   [-1, 1, Conv, [128, 3, 2]],  # 1-P2/4
   [-1, 3, C3, [128]]
  ],  I understand number, module, args,  what is the purpose of "from"? It is a little difficult for me to understand this from? 

2. anchors:  based on my understanding from anchor explanation from YOLOV3
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32
it means for each grid, there are 3 predefined anchor boxes for each grid [10,13,     16,30,     33,23] . But what is the unit of these number, it is pixel number or what?  It is a little confusing for me.  Your explanation will help me better understand your code.

**Comment by frabob2017** [comment]
No one answer this question?

**Comment by frabob2017** [comment]
I noticed C3TR,C3SPP, C3Ghost, these class derived from C3 class instead of nn.module,  I understand why there is no forward method is defined. But I still do not see where these classes are used in code?

**Comment by pderrenger** [comment]
@frabob2017 thank you for your interest in understanding the YOLOv8 source code. 🌟 I'll be happy to help clarify a few key points.

Regarding your first question about the "from" parameter in the yolov5.yaml file, the "from" field indicates the source index of the preceding layer from which the current layer receives input. It helps to establish the input connection between layers and allows for more flexibility in defining the network topology.

As for the anchors in the YOLOv8 architecture, the numbers provided represent the width and height of the anchor boxes relative to the input image size. These anchor box dimensions are in pixels and are used to detect objects at different scales within the image.

Lastly, the classes C3TR, C3SPP, and C3Ghost are indeed derived from the C3 class, but they are not used directly in the YOLOv8 implementation. They serve as template classes to encapsulate specific layer configurations for potential future use or experimentation.

I hope this helps clarify the areas you were inquiring about. 🚀 If you have further questions or need additional assistance, please don't hesitate to ask.

**Comment by glenn-jocher** [comment]
@pderrenger, I appreciate your engagement with the YOLOv8 source code and your efforts to adapt it for medical imaging. 👍

1. The "from" parameter in the model configuration files (like yolov5.yaml) is used to define the input layers for each module. It specifies which layer outputs should be used as inputs for the current layer, allowing for complex architectures with skip connections and concatenations.

2. The anchor values represent the width and height of predefined bounding boxes in pixels, scaled to the model's input image size. These anchors help the model predict object sizes more accurately.

Regarding the classes like C3TR, C3SPP, and C3Ghost, while they inherit from the C3 class, they are not utilized in the standard YOLOv8 model. They are part of the codebase for potential extensions or custom modifications and may not have a forward method defined if they rely on the parent class's implementation.

I'm glad to see your enthusiasm for extending YOLOv8, and I encourage you to explore the documentation for further guidance. Keep up the great work! 🌐🔍
